### PR TITLE
Add native mineflayer support for particles

### LINKED ISSUE #1887: Add a native mineflayer event for particles
## Is your feature request related to a problem? Please describe.
Shouldn't have to rely on node-minecraft-protocol for particle packets

## Describe the solution you'd like
`particles` event

## Describe alternatives you've considered
using `bot._client.once('world_particles', onParticles)`

## Additional context
N/A

### PR DESCRIPTION
**Pull request by Zn10plays**
Closes #1887

### CODE REVIEW

**Comment by Zn10plays** [comment]
ready for review

File: docs/api.md [hunk 1]
```diff
@@ -951,6 +952,11 @@ Setting values for this object internally calls [bot.setControlState](#botsetcon
 
 ### Events
 
+#### "particle" (particle)
+
+Only emitted when a particle is created.
+ * `particle` - The particle created, type Json
```
Comment (u9g) [inline docs/api.md:958]: this gives no info

File: lib/plugins/particles.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+module.exports = inject
+
+function inject (bot) {
+  bot._client.on('world_particles', (packet) => {
+    bot.emit('particle', packet)
```
Comment (u9g) [inline lib/plugins/particles.js:5]: we dont want to give the whole packet, just the data

**Comment by u9g** [comment]
wouldn't the plural "particles" be better than the singular "particle"?

File: docs/api.md [hunk 3]
```diff
@@ -124,6 +124,7 @@
       - [bot.scoreboard](#botscoreboard)
       - [bot.controlState](#botcontrolstate)
     - [Events](#events)
+      - ["particle" (particle)](#particle-particle)
```
Comment (u9g) [inline docs/api.md:127]: it says not to edit the toc at the top, we have a tool hat makes the toc

File: docs/api.md (hunk 1 above)
Comment (Zn10plays) [reply docs/api.md:958]: then what should I write?

the paket shows ex.
```json
longDistance: false,
  x: -64,
  y: 68,
  offsetX: 8,
  offsetY: 8,
  offsetZ: 8,
  particleData: 16,
  particles: 9000,
  data: undefined
} { size: 50, name: 'world_particles', state: 'play' }
```

File: lib/plugins/particles.js (hunk 2 above)
Comment (Zn10plays) [reply lib/plugins/particles.js:5]: but data usually is undefined (if we are strictly peaking what the event emits)

File: docs/api.md (hunk 3 above)
Comment (Zn10plays) [reply docs/api.md:127]: I will uncommit it

File: docs/api.md (hunk 1 above)
Comment (u9g) [reply docs/api.md:958]: this isnt a valid js object, what does the actual object look like?

Comment (Zn10plays) [reply docs/api.md:958]: Exactly what I typed

**Comment by Zn10plays** [comment]
also changed from singular to plural, 
changed the event name from 'particle to 'particles'

File: docs/api.md [hunk 4]
```diff
@@ -951,6 +951,22 @@ Setting values for this object internally calls [bot.setControlState](#botsetcon
 
 ### Events
 
+#### "particles" (particle)
+
+Only emitted when a particle is created.
+Particle is a Json object that contains:
+ * `particleId` - The particle Id.
```
Comment (extremeheat) [inline docs/api.md:958]: Abstract and document this

File: docs/api.md [hunk 5]
```diff
@@ -951,6 +951,22 @@ Setting values for this object internally calls [bot.setControlState](#botsetcon
...
+ * `longDistance` - Boolen, if the particle is supper far
+ * `x` - The particle's x location
+ * `y` - The particle's y location
+ * `z` - The particle's z location
+ * `offsetX` - This is added to the X position after being multiplied by random.nextGaussian()
+ * `offsetY` - This is added to the Y position after being multiplied by random.nextGaussian()
+ * `offsetZ` - This is added to the Z position after being multiplied by random.nextGaussian()
+ * `particleData` - The data of each particle.
```
Comment (extremeheat) [inline docs/api.md:966]: What data?

**Review by extremeheat — CHANGES_REQUESTED** [review]
For the public facing API, you need to abstract the protocol away more. Otherwise for every update we have to maintain the current schema and that could mean breaking changes in the future. Look at how boss bar is implemented currently for example.

File: lib/plugins/particles.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+module.exports = inject
+
+function inject (bot) {
+  bot._client.on('world_particles', (data) => {
+    bot.emit('particles', data)
```
Comment (rom1504) [inline lib/plugins/particles.js:5]: that's not good as @extremeheat said, there is no abstraction layer here, it is very likely already broken between versions

Comment (KaffinPX) [reply lib/plugins/particles.js:5]: In 1.8 - 1.16 only packets are changed not packets name, maybe parsing it for each version in same format will good idea.

File: lib/plugins/particles.js [hunk 7]
```diff
@@ -1,7 +1,52 @@
...
+
+  get offset () {
+    return this._offset
+  }
+}
+
 function inject (bot) {
   bot._client.on('world_particles', (data) => {
-    bot.emit('particles', data)
+    console.log(data)
```
Comment (u9g) [inline lib/plugins/particles.js:48]: don't forget to remove this

File: lib/plugins/particles.js [hunk 8]
```diff
@@ -1,7 +1,52 @@
...
+    const particle = new Particle(data.particleId, data.particles, [data.x, data.y, data.z], [data.offsetX, data.offsetY, data.offsetZ])
```
[the remark this answers is not in the document]
Comment (KaffinPX) [reply lib/plugins/particles.js:49]: ^^^

**Comment by rom1504** [comment]
is this ready ?

**Comment by Zn10plays** [comment]
Not ready still need to finish on the abstraction for the data property

**Comment by Zn10plays** [comment]
I am going to delete this pr because I feel like I will never get back to it. Though I WILL work on somthing else.

**Pull request closed without merging**